pkg/cmd/cli/cmd/login.go

```diff
@@ -0,0 +1,250 @@
+	kclientcmd "github.com/GoogleCloudPlatform/kubernetes/pkg/client/clientcmd"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+
+	"github.com/openshift/origin/pkg/client"
+	"github.com/openshift/origin/pkg/cmd/cli/config"
+	"github.com/openshift/origin/pkg/cmd/util"
+	osclientcmd "github.com/openshift/origin/pkg/cmd/util/clientcmd"
+	"github.com/openshift/origin/pkg/cmd/util/tokencmd"
+)
+
+type loginOptions struct {
+	username string
+	password string
+	server   string
+	project  string
+	context  string
+}
+
+func NewCmdLogin(f *osclientcmd.Factory, out io.Writer) *cobra.Command {
+	options := &loginOptions{}
+
+	cmds := &cobra.Command{
+		Use:   "login [--username=<username>] [--password=<password>] [--server=<server>]",
+		Short: "Logs in and returns a session token",
+		Long: `Logs in to the OpenShift server and prints out a session token.
+
+Username and password can be provided through flags, the command will
+prompt for user input if not provided.
+`,
+		Run: func(cmd *cobra.Command, args []string) {
+			// fetch config and if a server were not provided, prompt for it
+			clientCfg, err := f.OpenShiftClientConfig.ClientConfig()
+			if isServerNotFound(err) {
+				defaultServer := osclientcmd.DefaultClusterURL
+				promptedServer := util.PromptForStringWithDefault(os.Stdin, defaultServer, "OpenShift is not configured. Please provide the server URL or hit <enter> to use '%v': ", defaultServer)
+				serverFlag := cmd.Flags().Lookup("server")
+				if serverFlag != nil {
+					serverFlag.Value.Set(promptedServer)
+					if err := cmd.Execute(); err != nil {
+						os.Exit(1)
+					} else {
+						os.Exit(0)
+					}
+				}
+			}
+			checkErr(err)
+
+			// try to find a config file from default locations, or create a new one if not found
+			configStore, err := config.GetConfigFromDefaultLocations(clientCfg, cmd)
+			checkErr(err)
+
+			// validate flags
+			checkErr(options.validate(configStore, clientCfg))
+
+			// Check to see if we're already signed in (with the user provided by --username)
+			// If so, simply make sure that .kubeconfig has that information
+			if userFullName, err := whoami(clientCfg); err == nil && (!options.usernameProvided() || (options.usernameProvided() && options.username == userFullName)) {
+				err = config.UpdateConfigFile(userFullName, clientCfg.BearerToken, f.OpenShiftClientConfig.KClientConfig, configStore)
+				checkErr(err)
+				fmt.Printf("Already logged into %v as '%v'\n", clientCfg.Host, userFullName)
+
+			} else {
+				requiresNewLogin := true
+
+				// if trying to log in an user that's not the currently logged in, try to reuse a previous token
+				if options.usernameProvided() {
+					glog.V(5).Infof("Checking existing authentication info for '%v'...\n", options.username)
+
+					for key, authInfo := range configStore.Config.AuthInfos {
+						glog.V(5).Infof("Checking %v...\n", key)
+
+						if key == options.username && len(authInfo.Token) > 0 { // only token for now
```

Comment:
Alternatively, if you're trying to go through all the tokens and find one that matches the provided username, check all of them.